### PR TITLE
Partial ShaderMask for Web DOM

### PR DESCRIPTION
**Pull request by anirudhb**
## Description

Adds a partial ShaderMask implementation on Flutter Web with the DOM backend.

What's implemented:
- [ ] `ui.Gradient` shaders
  - [x] Linear gradients
  - [x] Radial gradients
  - [ ] Sweep gradients
  - [ ] Conical gradients
- [ ] `ui.ImageShader` shaders

All blend modes should work.

## Related Issues

flutter/flutter#44152

## Tests

I added the following tests:
 - None, this is not breaking behavior.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

**Comment by anirudhb** [comment]
I should be able to add support for the rest of the color modes using SVG filters as well, so I'll work on that too.  Regarding the rest of the shaders, image shaders should be possible with background-repeat but sweep gradients might be hard to implement (not sure how to implement those.)

File: lib/web_ui/lib/src/engine/html/shader_mask.dart [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+//
+// R' = r1*R + r2*G + r3*B + r4*A + r5
+// G' = g1*R + g2*G + g3*B + g4*A + g5
+// B' = b1*R + b2*G + b3*B + b4*A + b5
+// A' = a1*R + a2*G + a3*B + a4*A + a5
+String _srcInImageToSvg(String imageUrl) {
+  _maskFilterIdCounter += 1;
+  return '<svg width="0" height="0" xmlns:xlink="http://www.w3.org/1999/xlink">'
+      '<filter id="_fmf$_maskFilterIdCounter" '
+      'filterUnits="objectBoundingBox">'
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/shader_mask.dart:295]: `objectBoundingBox` is the default. We don't need to specify it. Will save a bit of CPU/memory/network.

(applies here and in other filters)

File: lib/web_ui/lib/src/engine/html/shader_mask.dart [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+    case ui.BlendMode.dstATop:
+    case ui.BlendMode.dstIn:
+    case ui.BlendMode.dstOut:
+    case ui.BlendMode.dstOver:
+    case ui.BlendMode.clear:
+    case ui.BlendMode.srcOver:
+      assert(
+          false,
+          'Invalid svg filter request for blend-mode '
+          '$blendMode');
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/shader_mask.dart:270]: Instead of `assert` followed by `break` let's just `throw UnsupportedError`.

**Review by yjbanov — COMMENTED** [review]
Thanks for contributing the web implementation! I have some preliminary comments. @ferhatb will probably have more. We can't land it without tests. For this functionality specifically, we'll need screenshot tests. I think we can cover all of the functionality with a single screenshot ([example](https://github.com/flutter/engine/blob/d2ea378be45dd0de7a86b521d48240fd9d276283/lib/web_ui/test/golden_tests/engine/canvas_mask_filter_test.dart#L54)).

File: lib/web_ui/lib/src/engine/html/shader_mask.dart [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+      return;
+    }
+    final gradient = shader as ui.Gradient;
+    // TODO(flutter-web): Support conical and sweep gradients for ShaderMask.
+    if (gradient is GradientConical || gradient is GradientSweep) {
+      return;
+    }
+
+    String inputImageSvg;
+    if (gradient is GradientLinear) {
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/html/shader_mask.dart:77]: would be nice to refactor this into a utility function :  _gradientToSvgElement(EngineGradient gradient...) , so in the error case you can throw ArgumentError.

File: lib/web_ui/lib/src/engine/html/shader_mask.dart [hunk 4]
```diff
@@ -0,0 +1,376 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/html/shader_mask.dart:1]: Needs golden test to verify it renders on both Chrome and WebKit.

File: lib/web_ui/lib/src/engine/html/shader_mask.dart [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+      '</feImage>'
+      '<feComposite in="image" in2="destalpha" '
+      'operator="arithmetic" k1="1" k2="0" k3="0" k4="0" result="comp">'
+      '</feComposite>'
+      '</filter></svg>';
+}
+
+String _srcOutImageToSvg(String imageUrl) {
+  _maskFilterIdCounter += 1;
+  return '<svg width="0" height="0" xmlns:xlink="http://www.w3.org/1999/xlink">'
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/html/shader_mask.dart:310]: Consider moving common header to small function.

**Review by ferhatb — CHANGES_REQUESTED** [review]

File: lib/web_ui/lib/src/engine/html/shader_mask.dart (hunk 1 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/shader_mask.dart:295]: Please test in Firefox. There are some cross browser bugs around filterUnits

**Comment by ferhatb** [comment]
Hi @anirudhb are you still working on this? Anyway we can help?

**Comment by anirudhb** [comment]
Just updated the PR. I'm currently writing golden tests, but not sure what the process is to generate/submit golden files (so that the test does not fail.)

**Comment by ferhatb** [comment]
> Just updated the PR. I'm currently writing golden tests, but not sure what the process is to generate/submit golden files (so that the test does not fail.)
> 
> Edit: It seems my most recent commit somehow messed up Firefox's rendering. Looking into and fixing. Although afaik the lack of `filterUnits` should not affect rendering.

Once you have the test

> Just updated the PR. I'm currently writing golden tests, but not sure what the process is to generate/submit golden files (so that the test does not fail.)
> 
> Edit: It seems my most recent commit somehow messed up Firefox's rendering. Looking into and fixing. Although afaik the lack of `filterUnits` should not affect rendering.

Great! Once you have firefox problem solved and add a test we can gen goldens.

- Please look at gradient_golden_test.dart and create a shadermask_golden_test.dart that is similar.
- engine/src/flutter/lib/web_ui>. dev/felt test test/engine/golden_tests/shadermask_golden_test.dart
- then if you have a mac , repeat the above with:  --browser=ios-safari
- Under the .dart_tool/golden/web directory you should now see the golden files
- Add those files to flutter/goldens repo in a PR and send out for review.
- Finally take the commit hash of that golden repo commit and update golden_locks.yaml with new commit hash in this PR.
- 
- If you don't have a mac to generate safari goldens, no worries, add the test, let us know and we'll generate goldens PR for you.

**Comment by yjbanov** [comment]
> * engine/src/flutter/lib/web_ui>. dev/felt test test/engine/golden_tests/shadermask_golden_test.dart
> * then if you have a mac , repeat the above with:  --browser=ios-safari
> * Under the .dart_tool/golden/web directory you should now see the golden files

In order for `felt` to generate the new goldens you also need to pass `--update-screenshot-goldens` (or pass `write: true` to `matchGoldenFile`). If you don't have access to a Mac or Linux, you can just write the test, mark it as `skip: true`, and post the screenshots generated on whatever computer you have here for manual review. We can enable them on the CI ourselves in a separate PR.

**Comment by yjbanov** [comment]
@anirudhb do you plan to continue working on this PR?

**Comment by anirudhb** [comment]
I'll take a look at it again this weekend.

**Comment by ferhatb** [comment]
@anirudhb , please take a look at https://github.com/flutter/engine/pull/24029 it contains a refactor of gradient code and your changes to support full set of gradients. Let me know if you have time to write the screenshot tests and add changes. If not we can add tests and land it.

**Comment by anirudhb** [comment]
I don't think I'll have time to fix up this PR, but feel free to make it usable for merge!

**Comment by MrBirb** [comment]
@anirudhb Looks the issue was closed. Is this PR still valid?

**Comment by ferhatb** [comment]
Closing this PR shadermasks have landed in a separate PR.

**Pull request closed without merging**